### PR TITLE
CO2 meter extension - a sample for using WebHID on extension service worker.

### PR DESCRIPTION
**Pull request by chengweih001**
Hi Chrome extensions samples maintainers,

This is Jack (chengweih@) from Chrome Device API team. I would like to submit this sample as a reference for developers for using WebHID on extension service workers.

The WebHID on extension service workers feature is now targeting to launch in Chrome 115 ([launch bug](https://launch.corp.google.com/launch/4200800?utm_source=1Alphabet%20Wide:Product%20Inclusion%20%26%20Equity:::Alphabet%20Wide:Privacy:::Alphabet%20Wide:Security:::Chrome%20Browser:Chrome%20Browser:::ChromeOS:ChromeOS%20Features&utm_medium=Email-notification-with-CTA&utm_campaign=Notification-APPROVER_UPDATE&utm_content=id:4200800-version:144)) with one limitation which is mentioned in the README.md file. The purpose of this sample is to convey how WebHID can be used in extension service workers and what the limitations are.

### CODE REVIEW

**Comment by sebastianbenz** [comment]
Thanks a lot for the PR Jack! The extension (and the code) look great. It looks like the extension is fully functional and does way more than simply demonstrating the features of WebHID. I'm a bit hesitant to add this to the Chrome Extension Samples repo as our samples are usually much simpler and more focused. 

Did you already consider moving your extension to it's own repository? The size and scope would definitely justify this imo. Wdyt?

**Comment by chengweih001** [comment]
Because this official Chrome extension repository is a popular place where developers come to learn how the API is used in extensions, I wish to have this extension sample available here to be explored.

If you prefer a simpler and more focused extension, I can try to remove some of the features of this extension and make it purely focused on using the WebHID API. For example, I will remove the storage and chart, and instead only show the current reading in the pop-up window.

If a simplified version is okay to you, I will revise this patch and update once ready.

**Comment by sebastianbenz** [comment]
I completely agree that we should have a WebHID sample here and a simpified version would be perfect! Thanks a lot for understanding.

**Comment by chengweih001** [comment]
I stripped down and made it a simpler version. Could you please take another look?

File: functional-samples/sample.co2meter/background.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    co2MeterConnected,
+    co2MeterDisconnected,
+    OnCO2Reading,
+    OnTempReading
+  );
+  startCO2Reading();
+}
+
+if (navigator.hid) {
+  initilize();
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/background.js:111]:
```suggestion
  initialize();
```

File: functional-samples/sample.co2meter/co2-state-iframe.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  NEW_CO2_READING,
+  NEW_TEMP_READING,
+  READING_UNKNOWN
+} from './modules/constant.js';
+import CO2Meter from './modules/co2_meter.js';
+
+window.onload = async () => {
+  // Register for messages to update chart upon new data readings.
+  chrome.runtime.connect().onMessage.addListener((msg) => {
+    if (msg.type === NEW_CO2_READING) {
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/co2-state-iframe.js:27]: maybe: a switch expression might be easier to read.

File: functional-samples/sample.co2meter/main-page.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import CO2Meter from './modules/co2_meter.js';
+
+window.onload = async () => {
+  // Permission
+  // Popup window can't open a permission prompt so we have to result to a page.
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/main-page.js:19]:
```suggestion
  // Popup window can't open a permission prompt so we have to use a page instead.
```

File: functional-samples/sample.co2meter/manifest.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  "manifest_version": 3,
+  "background": {
+    "service_worker": "background.js",
+    "type": "module"
+  },
+  "action": {
+    "default_icon": {
+      "32": "images/icon32.png"
+    },
+    "default_title": "Click me",
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/manifest.json:14]: Maybe use a more descriptive title

File: functional-samples/sample.co2meter/modules/co2_meter.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    }
+  }
+
+  /**
+   * @description Request user to grant permission for using CO2 meter.
+   * The extension currently only support this model:
+   * https://www.co2meter.com/products/co2mini-co2-indoor-air-quality-monitor
+   */
+  requestPermission() {
+    navigator.hid
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/modules/co2_meter.js:123]: Could we use async/await here for consistency?

File: functional-samples/sample.co2meter/manifest.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    "default_popup": "popup.html"
+  },
+  "permissions": ["scripting","alarms"],
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/manifest.json:17]: Do you need the alarms and scripting permissions?

File: functional-samples/sample.co2meter/manifest.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  "host_permissions": ["<all_urls>"],
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/manifest.json:18]: Why is this needed?

File: functional-samples/sample.co2meter/README.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+The extension uses [WebHID](https://developer.chrome.com/en/articles/hid/) to access a device for measuring the CO₂ level and temperature in your surroundings.
+
+## **Testing the extension**
+
+1. Follow the instructions to load an [unpacked extension](https://developer.chrome.com/docs/extensions/mv3/getstarted/development-basics/#load-unpacked).
+2. Connect the CO₂ meter (currently it only supports the [CO2Mini Indoor Air Quality Monitor](https://www.co2meter.com/products/co2mini-co2-indoor-air-quality-monitor) from CO2Meter.com).
+3. Open the extension popup window and click “Settings” button to go to the settings page.
+4. Click the “Grant CO2 meter permission” button and grant the permission to the CO₂ meter.
+
+Following the above steps, the device connection session to the CO₂ meter will be created when the extension is running. The input reports from the device will be processed and are visble from the popup window or settings page.
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/README.md:12]: I think it'd be great to have a section here that calls out and links to important parts of the implementation (e.g. ask for permission, list devices, connect to a device, ...)

File: functional-samples/sample.co2meter/main-page.html [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    <script type="module" src="main-page.js"></script>
+    <style>
+      iframe {
+        border: none;
+      }
+    </style>
+  </head>
+  <body>
+    <div>
+      <button type="button" id="grantPermissionButton">
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/main-page.html:13]: Would it be easy to hide this button if the permission has already been granted?

**Review by sebastianbenz — COMMENTED** [review]
This looks great - thanks for reworking! Left a few comments, but mostly nits.

File: functional-samples/sample.co2meter/background.js (hunk 1 above)
Comment (chengweih001) [reply functional-samples/sample.co2meter/background.js]: Done.

File: functional-samples/sample.co2meter/co2-state-iframe.js (hunk 2 above)
Comment (chengweih001) [reply functional-samples/sample.co2meter/co2-state-iframe.js]: Done.

File: functional-samples/sample.co2meter/main-page.js (hunk 3 above)
Comment (chengweih001) [reply functional-samples/sample.co2meter/main-page.js]: Done.

File: functional-samples/sample.co2meter/manifest.json (hunk 4 above)
Comment (chengweih001) [reply functional-samples/sample.co2meter/manifest.json]: Done.

File: functional-samples/sample.co2meter/modules/co2_meter.js (hunk 5 above)
Comment (chengweih001) [reply functional-samples/sample.co2meter/modules/co2_meter.js]: Done.

File: functional-samples/sample.co2meter/manifest.json (hunk 6 above)
Comment (chengweih001) [reply functional-samples/sample.co2meter/manifest.json]: No, these are left-over, cleared.

File: functional-samples/sample.co2meter/manifest.json (hunk 7 above)
Comment (chengweih001) [reply functional-samples/sample.co2meter/manifest.json]: This is left-over, cleared.

File: functional-samples/sample.co2meter/README.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+## **Design**
+
+- `co2_meter.js`: A CO2 meter device driver layer that uses WebHID to communicate with the device.
```
Comment (sebastianbenz) [inline functional-samples/sample.co2meter/README.md:16]: Can you link to the respective files?

**Review by sebastianbenz — APPROVED** [review]
Thanks a lot! One small nit otherwise LGTM.

File: functional-samples/sample.co2meter/main-page.html (hunk 9 above)
Comment (chengweih001) [reply functional-samples/sample.co2meter/main-page.html:13]: We don't have any plan to integrate that yet and. [navigator.hid.getDevices](https://wicg.github.io/webhid/#getdevices-method) does document that this `getDevices` method can be used for retrieving permitted devices.

File: functional-samples/sample.co2meter/README.md (hunk 10 above)
Comment (chengweih001) [reply functional-samples/sample.co2meter/README.md]: Done.

File: functional-samples/sample.co2meter/background.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  last_co2_reading = co2_reading;
+  await broadcastMessage(NEW_CO2_READING, co2_reading);
+}
+
+async function OnTempReading(temp_reading) {
+  last_temp_reading = temp_reading;
+  await broadcastMessage(NEW_TEMP_READING, temp_reading);
+}
+
+async function initilize() {
```
[the remark this answers is not in the document]
Comment (chengweih001) [reply functional-samples/sample.co2meter/background.js]: Done.

**Comment by sebastianbenz** [comment]
Thanks a lot @chengweih001! Great sample!
